service/forward_service.cc

```diff
@@ -412,27 +398,50 @@ future<query::forward_result> forward_service::execute_on_this_shard(
         cql3::query_options::specific_options::DEFAULT,
         cql_serialization_format::latest()
     );
-    auto pager = service::pager::query_pagers::pager(
-        _proxy,
-        schema,
-        selection,
-        *query_state,
-        *query_options,
-        make_lw_shared<query::read_command>(std::move(req.cmd)),
-        retain_ranges_owned_by_this_shard(schema, std::move(req.pr)),
-        nullptr // No filtering restrictions
-    );
+
     auto rs_builder = cql3::selection::result_set_builder(
         *selection,
         now,
         cql_serialization_format::latest(),
         std::vector<size_t>() // Represents empty GROUP BY indices.
     );
 
-    // Execute query.
-    while (!pager->is_exhausted()) {
-        co_await pager->fetch_page(rs_builder, DEFAULT_INTERNAL_PAGING_SIZE, now, timeout);
-    }
+    dht::partition_range_vector ranges_owned_by_this_shard;
+    partition_ranges_owned_by_this_shard owned_iter(schema, std::move(req.pr));
+
+    std::optional<dht::partition_range> current_range;
+    do {
+        // We serve up to 256 ranges at a time to avoid allocating a huge vector for ranges
+        static constexpr size_t max_ranges = 256;
+        while ((current_range = owned_iter.next(*schema))) {
+            ranges_owned_by_this_shard.push_back(*current_range);
+            if (ranges_owned_by_this_shard.size() >= max_ranges) {
+                break;
+            }
+        }
+        if (ranges_owned_by_this_shard.empty()) {
+            break;
+        }
+        flogger.trace("Forwarding to {} ranges owned by this shard", ranges_owned_by_this_shard.size());
+
+        auto pager = service::pager::query_pagers::pager(
+            _proxy,
+            schema,
+            selection,
+            *query_state,
+            *query_options,
+            make_lw_shared<query::read_command>(req.cmd),
+            std::move(ranges_owned_by_this_shard),
+            nullptr // No filtering restrictions
+        );
+
+        // Execute query.
+        while (!pager->is_exhausted()) {
+            co_await pager->fetch_page(rs_builder, DEFAULT_INTERNAL_PAGING_SIZE, now, timeout);
+        }
```

Comment:
Shouldn't you be merging the results of rs_builder into each other?

Reply:
That is, every cycle, extract a row from rs_builder, and merge it into an accumulator? And have a fresh rs_builder for each run, I don't know if it's designed for reuse.

Reply:
There is only one rs_builder, filled with multiple pagers. So it serves as its own accumulator here

Reply:
But won't it get multiple rows as a result?

Reply:
Each pager fetches data from disjoint ranges, so I think it would work exactly as before. If we used to have a vector of 4096 ranges [r0, r2, ... r4095], all this code does now is:
1. Use a pager to populate rs_builder with ranges [r0, r1, ..., r255]
...
16. Use a pager to populate rs_builder with ranges [r4080, ..., r4095]

Unless I severely misunderstood what this code, and `partition_ranges_owned_by_this_shard`, does

Reply:
s/complete/finalize/

Reply:
Ah, alright, I get the question. That the pager doesn't do any special finalizing action when exhausted, the only logic is in fetch_page, which blindly feeds rows into the result builder. The finalization is only performed in rs_builder.build()

Reply:
Thanks for clarifying, I'll merge.